src/dbnode/storage/index.go

```diff
@@ -715,80 +730,298 @@ func (i *nsIndex) Query(
+			ctx.RegisterFinalizer(blockResults)
+			// No merged results yet, use this as the first to merge into.
+			results.merged = blockResults
+		} else {
+			// Append the block results.
+			mergedResult = true
+			size := results.merged.Size()
+			for _, entry := range blockResults.Map().Iter() {
+				// Break early if reached limit.
+				if opts.Limit > 0 && size >= opts.Limit {
+					blockExhaustive = false
+					break
+				}
+
+				// Append to merged results.
+				id, tags := entry.Key(), entry.Value()
+				_, size, err = results.merged.AddIDAndTags(id, tags)
+				if err != nil {
+					results.multiErr = results.multiErr.Add(err)
+					return
+				}
+			}
 		}
 
-		exhaustive, err = block.Query(query, opts, results)
-		if err != nil {
-			return index.QueryResults{}, err
+		// If block had more data but we stopped early, need to notify caller.
+		if blockExhaustive {
+			return
 		}
+		results.exhaustive = false
+	}
 
-		if !exhaustive {
-			// i.e. block had more data but we stopped early, we know
-			// we have hit the limit and don't need to query any more.
-			break
+	for _, block := range blocks {
+		// Capture block for async query execution below.
+		block := block
+
+		// Terminate early if we know we don't need any more results.
+		results.Lock()
+		mergedSize := 0
+		if results.merged != nil {
+			mergedSize = results.merged.Size()
+		}
+		alreadyNotExhaustive := opts.Limit > 0 && mergedSize >= opts.Limit
+		if alreadyNotExhaustive {
+			results.exhaustive = false
 		}
+		results.Unlock()
 
-		// terminate if queryRange doesn't need any more data
-		queryRange = queryRange.RemoveRange(blockRange)
-		if queryRange.IsEmpty() {
+		if alreadyNotExhaustive {
+			// Break out if already exhaustive.
 			break
 		}
+
+		if applyTimeout := timeout > 0; !applyTimeout {
+			// No timeout, just wait blockingly for a worker.
+			wg.Add(1)
+			i.queryWorkersPool.Go(func() {
+				execBlockQuery(block)
+				wg.Done()
+			})
+			continue
+		}
+
+		// Need to apply timeout to the blocking wait call for a worker.
+		var timedOut bool
+		if timeLeft := deadline.Sub(i.nowFn()); timeLeft > 0 {
+			wg.Add(1)
+			timedOut := !i.queryWorkersPool.GoWithTimeout(func() {
+				execBlockQuery(block)
+				wg.Done()
+			}, timeLeft)
+
+			if timedOut {
+				// Did not launch task, need to ensure don't wait for it.
+				wg.Done()
+			}
+		} else {
+			timedOut = true
+		}
+
+		if timedOut {
+			// Exceeded our deadline waiting for this block's query to start.
+			return index.QueryResults{}, fmt.Errorf("index query timed out: %s", timeout.String())
+		}
+	}
+
+	// Wait for queries to finish.
+	if !(timeout > 0) {
+		// No timeout, just blockingly wait.
+		wg.Wait()
+	} else {
+		// Need to abort early if timeout hit.
+		timeLeft := deadline.Sub(i.nowFn())
+		if timeLeft <= 0 {
+			return index.QueryResults{}, fmt.Errorf("index query timed out: %s", timeout.String())
+		}
+
+		var (
+			ticker  = time.NewTicker(timeLeft)
```

Comment:
I know Prateek has talked about time tickers not getting scheduled properly in certain situations and preferring to use loops with time.Sleep which apparently has some special hooks into the runtime, although then you'd have to implement polling so ehhhh. Probably fine as is